File: 백준/Gold/15686. 치킨 배달/치킨 배달.py
import sys, copy
from itertools import combinations
input = sys.stdin.readline

N, M = map(int, input().split())
city = []
house = []
chicken = []
answer = int(10e9)
for _ in range(N):
    city.append(list(map(int, input().split())))

for i in range(N):
    for j in range(N):
        if city[i][j] == 1:
            house.append((i,j))
        elif city[i][j] == 2:
            chicken.append((i,j))

def solution(select):
    global answer
    dist = 0
    for i in range(len(house)): # 집 갯수만큼
        result = int(10e9)
        hx, hy = house[i]
        for j in range(len(select)): # 집 1개당 가능한 치킨집 갯수만큼 계산
            cx, cy = select[j]
            result = min(result, (abs(hx - cx) + abs(hy - cy)))
        else:
            dist += result
    answer = min(answer,dist)

for pos in combinations(chicken, M):
    select = []
    for x, y in pos:
        select.append((x,y))
    solution(select)

print(answer)

File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.io.*;
import java.util.*;

public class Main {

	static class Node {
		int x;
		int y;

		public Node(final int x, final int y) {
			this.x = x;
			this.y = y;
		}
	}

	static ArrayList<Node> home = new ArrayList<>();
	static ArrayList<Node> chicken = new ArrayList<>();
	static int answer = Integer.MAX_VALUE;

	static int[] order;


	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st;

		st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());

		int[][] city = new int[N][N];
		order = new int[M];

		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());


			for (int j = 0; j < N; j++) {
				int value = Integer.parseInt(st.nextToken());
				city[i][j] = value;
				if (value == 1) {
					home.add(new Node(i, j));
				}

				if (value == 2) {
					chicken.add(new Node(i, j));
				}
			}
		}

		combination(0, M, 0);

		System.out.println(answer);
	}

	private static void combination(int depth, int M, int start) {

		if (depth == M) {
			answer = Math.min(answer, getDistance(order));
			return;
		}

		for (int i = start; i < chicken.size(); i++) {
			order[depth] = i;
			combination(depth + 1, M, i + 1);
		}
	}

	private static int getDistance(int[] order) {

		int sum = 0;

		for (int i = 0; i < home.size(); i++) {
			int result = Integer.MAX_VALUE;
			Node homePos = home.get(i);

			for(int o : order) {
				Node chickenPos = chicken.get(o);
				result = Math.min(result, Math.abs(homePos.x - chickenPos.x) + Math.abs(homePos.y - chickenPos.y));
			}

			sum += result;

		}

		return sum;
	}
}
